gms/gossiper.cc

```diff
@@ -340,10 +340,20 @@ future<> gossiper::do_send_ack2_msg(msg_addr from, utils::chunked_vector<gossip_
         std::map<inet_address, endpoint_state> delta_ep_state_map;
         for (auto g_digest : ack_msg_digest) {
             inet_address addr = g_digest.get_endpoint();
-            auto local_ep_state_ptr = this->get_state_for_version_bigger_than(addr, version_type(g_digest.get_max_version()));
-            if (local_ep_state_ptr) {
-                delta_ep_state_map.emplace(addr, *local_ep_state_ptr);
+            const auto es = get_endpoint_state_for_endpoint_ptr(addr);
+            if (!es || es->get_heart_beat_state().get_generation() < g_digest.get_generation()) {
+                continue;
             }
+            // Local generation for addr may have been increased since the
+            // current node sent an initial SYN. Comparing versions across
+            // different generations in get_state_for_version_bigger_than
+            // could result in loosing some app states with smaller versions.
+            const auto version = es->get_heart_beat_state().get_generation() > g_digest.get_generation()
+                ? version_type(0)
+                : g_digest.get_max_version();
+            auto local_ep_state_ptr = this->get_state_for_version_bigger_than(addr, version);
+            assert(local_ep_state_ptr);
```

Comment:
asserts are banned in favor of on_internal_error().

Why assert() and not an "if" like before?

Reply:
changed the code a bit, now there is
```
if (!es || es->get_heart_beat_state().get_generation() < g_digest.get_generation()) {
    continue;
}
```
above, `get_state_for_version_bigger_than` returns null only if it can't find the endpoint and this can't happen now.

Reply:
can change it to `on_internal_error` or revert to how it was before

Reply:
get_state_for_version_bigger_than() also returns an empty optional if there are no states which match the version filter and local_hb_version <= version

Reply:
> asserts are banned in favor of on_internal_error().
> 

Is it? Where is it written? If it is the rule is enforced inconsistently which is annoying. Here is very recent commit with an assert: 00db47292bdbd21e8fa72a6f681ee30e625b90bd. And here is relatively recent commit of yours: fe7922d65cbfaf924b6d6a9504cb588655d544bc

Reply:
> get_state_for_version_bigger_than() also returns an empty optional if there are no states which match the version filter and local_hb_version <= version

fixed

Reply:
> > asserts are banned in favor of on_internal_error().
> 
> Is it? Where is it written?

There was a discussion in https://github.com/scylladb/scylladb/issues/7871. I don't think it's documented yet.

> If it is the rule is enforced inconsistently which is annoying. Here is very recent commit with an assert: [00db472](https://github.com/scylladb/scylladb/commit/00db47292bdbd21e8fa72a6f681ee30e625b90bd). And here is relatively recent commit of yours: [fe7922d](https://github.com/scylladb/scylladb/commit/fe7922d65cbfaf924b6d6a9504cb588655d544bc)

I can see why it's annoying, we should be more consistent.